scripts/release/proposal.js

```diff
@@ -0,0 +1,128 @@
+].join(' ')
+
+// Determine the new version.
+const [lastMajor, lastMinor, lastPatch] = require('../../package.json').version.split('.').map(Number)
+const lineDiff = capture(`${diffCmd} v${releaseLine}.x master`)
+const newVersion = lineDiff.includes('SEMVER-MINOR')
+  ? `${releaseLine}.${lastMinor + 1}.0`
+  : `${releaseLine}.${lastMinor}.${lastPatch + 1}`
+
+// Checkout new branch and output new changes.
+run(`git checkout v${newVersion}-proposal || git checkout -b v${newVersion}-proposal`)
+
+// Get the hashes of the last version and the commits to add.
+const lastCommit = capture('git log -1 --pretty=%B').trim()
+const proposalDiff = capture(`${diffCmd} --format=sha --reverse v${newVersion}-proposal master`)
+  .replace(/\n/g, ' ').trim()
+
+if (proposalDiff) {
+  // We have new commits to add, so revert the version commit if it exists.
+  if (lastCommit === `v${newVersion}`) {
+    run('git reset --hard HEAD~1')
+  }
+
+  // Output new changes since last commit of the proposal branch.
+  run(`${diffCmd} v${newVersion}-proposal master`)
+
+  // Cherry pick all new commits to the proposal branch.
+  try {
+    run(`echo "${proposalDiff}" | xargs git cherry-pick`)
+  } catch (err) {
+    error('Cherry-pick failed. Resolve the conflicts and run `git cherry-pick --continue` to continue.')
+    error('When all conflicts have been resolved, run this script again.')
+    process.exit(1)
+  }
+}
+
+// Update package.json with new version.
+run(`yarn version --no-git-tag-version --new-version ${newVersion}`)
+run(`git commit -uno -m v${newVersion} package.json || exit 0`)
+
+ready()
+
+// Check if current branch is already an active patch proposal branch to avoid
+// creating a new minor proposal branch if new minor commits are added to the
+// main branch during a existing patch release.
+function isActivePatch () {
+  const currentMatch = currentBranch.match(/^(\d+)\.(\d+)\.(\d+)-proposal$/)
+
+  if (currentMatch) {
+    const [major, minor, patch] = currentMatch.slice(1).map(Number)
+
+    if (major === lastMajor && minor === lastMinor && patch > lastPatch) {
+      return true
+    }
+  }
+
+  return false
+}
+
+// Output a command to the terminal and execute it.
+function run (cmd) {
+  whisper(`> ${cmd}`)
+
+  const output = execSync(cmd, {}).toString()
+
+  log(output)
+}
+
+// Run a command and capture its output to return it to the caller.
+function capture (cmd) {
+  return execSync(cmd, {}).toString()
+}
+
+// Write release notes to a file that can be copied to the GitHub release.
+function ready () {
+  const notesDir = path.join(__dirname, '..', '..', '.github', 'notes')
+  const notesFile = path.join(notesDir, `${newVersion}.md`)
+  const lineDiff = capture(`${diffCmd} --markdown=true v${releaseLine}.x master`)
+
+  fs.mkdirSync(notesDir, { recursive: true })
+  fs.writeFileSync(notesFile, lineDiff)
+
+  success('Release proposal is ready.')
+  success(`Changelog at .github/notes/${newVersion}.md`)
```

Comment:
I'm not a huge fan of leaving files around when we can avoid it.

My suggestion here would be to use [`git notes`](https://git-scm.com/docs/git-notes). While they don't show up in the UI (and haven't in 10 years! (_shakes fist at GitHub_)), They're a handy place to stash information that's preserved in git, and associated with a commit, so we can use that to store the release notes and if we want to automate even further, have a GitHub Action that copies the data from `git notes` into the PR description.

Alternatively... just log it instead of writing it to a file.

Reply:
Logging it means it could be lost, and once the release is done it becomes more complicated to generate it as branch-diff would no longer find the same commits. I don't see a problem with leftover files in a gitignored folder to be honest. A lot of tools leave files around and you can just delete the folder if you want them removed. Using git notes makes this more akward and you need to check how to use it to be able to extract the release notes from it. Opening a file and copy pasting its content is a lot easier. Eventually we could also use the files as an actual changelog, or do this automatically, but for now it seemed like the easiest to use solution.